cpp/src/groupby/sort/groupby_kernels.cuh

```diff
@@ -0,0 +1,271 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef GROUPBY_KERNELS_CUH
+#define GROUPBY_KERNELS_CUH
+
+#include <cudf/groupby.hpp>
+#include "../common/type_info.hpp"
+#include "utilities/device_atomics.cuh"
+
+using namespace cudf::groupby::common;
+
+namespace cudf {
+namespace groupby {
+namespace sort {
+/**---------------------------------------------------------------------------*
+ * @brief Base case for invalid SourceType and op combinations.
+ *
+ * For an invalid combination of SourceType and operator,
+ *`target_type_t<SourceType, operator>` yields a `void` TargetType. This
+ * specialization will be invoked for any invalid combination and cause a
+ * runtime error.
+ *
+ * @note A struct is used instead of a function to allow for partial
+ * specialization.
+ *---------------------------------------------------------------------------**/
+template <typename SourceType, operators op, bool values_have_nulls,
+          typename Enable = void>
+struct update_target_element {
```

Comment:
I assume you plan to move all of the redundant `device` functions shared between the sort/hash groupby into a single shared file?

Reply:
Yes. In order to do  that and  have different namespaces for sort-based groupby and hash based groupby I will insert  a new template parameter like `typename OperatorType` for `operators` enum.  Do you agree with that?

Reply:
I'm not following. Why do the operators need a template type?

Reply:
The problem is that many of the code that I put in the `common` folder depends on `operators` enum type.  That why in this first approach I use a unique enum type for both groupbys. 

So, in order to not duplicate code for functions that depends on operators type, we could  add this type enum type as a template parameter and use enable_if if there is specific behavior for any kind of groupby.    

What do you think?